File: day1-kotlin-syntax-crash-course/java/NullSafetyExample.java
/**
 * Day 1 - Null Safety in Java (compared to Kotlin)
 *
 * Java approach:
 *   - Any reference can be null (except primitives)
 *   - NullPointerException is the #1 runtime exception
 *   - No compile-time null checks
 *   - Workarounds: Optional, @Nullable/@NonNull annotations, manual checks
 *
 * Kotlin approach:
 *   - Null safety is BUILT INTO the type system
 *   - Types are non-null by default
 *   - Nullable types explicitly marked with ?
 *   - Safe operators: ?. !! ?: let
 */
import java.util.Optional;

public class NullSafetyExample {

    // ═══════════════════════════════════════════════════════════════
    // 1. JAVA: Any reference CAN be null
    // ═══════════════════════════════════════════════════════════════

    private static String name = null;     // ✅ OK in Java
    private static String name2 = "Hello"; // Also OK

    // ═══════════════════════════════════════════════════════════════
    // 2. JAVA NULL CHECKS (Manual — error-prone)
    // ═══════════════════════════════════════════════════════════════

    public static int getStringLength(String s) {
        // Java: You MUST manually check for null
        if (s != null) {
            return s.length();
        }
        return 0;  // or throw exception
    }

    // ═══════════════════════════════════════════════════════════════
    // 3. JAVA: Safe navigation is CLUMSY
    // ═══════════════════════════════════════════════════════════════

    // Nested null check (ugly!)
    public static String getCityName(Object address) {
        if (address instanceof java.util.Map) {
            @SuppressWarnings("unchecked")
            java.util.Map<String, Object> map = (java.util.Map<String, Object>) address;
            Object city = map.get("city");
            if (city instanceof String) {
                return (String) city;
            }
        }
        return "Unknown";
    }

    // ═══════════════════════════════════════════════════════════════
    // 4. JAVA: Optional<T> (Java 8+ — like Kotlin's nullable type)
    // ═══════════════════════════════════════════════════════════════

    public static Optional<String> findName(int id) {
        if (id == 1) {
            return Optional.of("Alice");
        }
        return Optional.empty();  // Instead of null
    }

    // ═══════════════════════════════════════════════════════════════
    // 5. JAVA: @Nullable / @NonNull annotations (Static analysis)
    // ═══════════════════════════════════════════════════════════════

    // Annotations don't prevent null at runtime, only help IDE/static analysis
    // public @NonNull String getName() { ... }
    // public @Nullable String findUser(int id) { ... }

    // ═══════════════════════════════════════════════════════════════
    // 6. COMPARISON TABLE
    // ═══════════════════════════════════════════════════════════════
    //
    //   Kotlin                      Java
    //   ─────────────────────────── ──────────────────────────────
    //   var name: String = "Hi"     String name = "Hi"
    //   var name: String? = null    String name = null
    //   name?.length                name != null ? name.length() : null
    //   name!!.length               name.length()  (crashes if null!)
    //   name ?: "default"           name != null ? name : "default"
    //   name?.let { ... }           if (name != null) { ... }
    //
    // ═══════════════════════════════════════════════════════════════

    public static void main(String[] args) {

        // ═══════════════════════════════════════════════════════════════
        // 7. JAVA NULL HANDLING PATTERNS
        // ═══════════════════════════════════════════════════════════════

        // Pattern 1: Manual null check
        String name = null;
        if (name != null) {
            System.out.println("Name length: " + name.length());
        } else {
            System.out.println("Name is null");
        }

        // Pattern 2: Ternary operator
        int length = (name != null) ? name.length() : 0;

        // Pattern 3: Optional (Java 8+)
        Optional<String> optName = Optional.ofNullable(name);
        int len = optName.map(String::length).orElse(0);

        // Pattern 4: Objects utility class
        int len2 = java.util.Objects.requireNonNullElse(name, "").length();

        // Pattern 5: Try-catch for NPE (ugly!)
        try {
            int len3 = name.length();  // Will throw NPE!
        } catch (NullPointerException e) {
            System.out.println("Name was null!");
        }

        // ═══════════════════════════════════════════════════════════════
        // 8. JAVA STREAMS + Optional (closest to Kotlin's approach)
        // ═══════════════════════════════════════════════════════════════

        Optional.ofNullable(name)
            .filter(n -> !n.isEmpty())
            .ifPresent(n -> System.out.println("Name: " + n));

        // ═══════════════════════════════════════════════════════════════
        // 9. KEY TAKEAWAYS FOR JAVA → KOTLIN
        // ═══════════════════════════════════════════════════════════════
        //
        // JAVA PROBLEMS:
        //   1. NullPointerExceptions at runtime (not compile-time)
        //   2. Manual null checks everywhere (verbose, error-prone)
        //   3. No language-level distinction between nullable/non-nullable
        //   4. Optional is for return types only, not fields
        //
        // KOTLIN SOLUTIONS:
        //   1. Nullable types (String?) are explicit in the type
        //   2. Safe call operator (?.) eliminates manual checks
        //   3. The compiler ENFORCEs null checks
        //   4. !! operator: explicit "I know this is not null" (rare)
        //   5. ?: operator: Elvis operator for defaults
        //   6. let: scoped null checks
        //
        // ═══════════════════════════════════════════════════════════════
    }
}

File: day1-kotlin-syntax-crash-course/kotlin/NullSafetyExample.kt
/**
 * Day 1 - Null Safety in Kotlin
 *
 * Kotlin's null safety is one of its biggest advantages over Java:
 *   - Types are NON-NULL by default
 *   - Nullable types marked with ? (e.g., String?)
 *   - Safe call operator: ?.
 *   - Non-null assertion: !!
 *   - Elvis operator: ?:
 *   - Let function for scoped operations
 */

fun main() {

    // ═══════════════════════════════════════════════════════════════
    // 1. NON-NULL TYPES (default in Kotlin)
    // ═══════════════════════════════════════════════════════════════

    val name: String = "Kotlin"
    // val nullName: String = null  // ❌ COMPILE ERROR: null cannot be a String

    // ═══════════════════════════════════════════════════════════════
    // 2. NULLABLE TYPES (explicit with ?)
    // ═══════════════════════════════════════════════════════════════

    val nullableName: String? = null   // ✅ OK — String? can be null
    val nonNullName: String = "Kotlin" // ✅ OK

    println("nullableName: $nullableName")
    println("nonNullName: $nonNullName")

    // ═══════════════════════════════════════════════════════════════
    // 3. SAFE CALL OPERATOR ?.
    // ═══════════════════════════════════════════════════════════════

    // If nullableName is null, expression returns null (no NPE!)
    val length1: Int? = nullableName?.length
    println("length1 (null): $length1")  // null

    val nonNull: String? = "Hello"
    val length2: Int? = nonNull?.length
    println("length2 (non-null): $length2")  // 5

    // Chained safe calls
    data class Address(val city: String?)
    data class Person(val address: Address?)

    val person: Person? = Person(Address("Hanoi"))
    val city: String? = person?.address?.city
    println("city: $city")  // Hanoi

    // ═══════════════════════════════════════════════════════════════
    // 4. NON-NULL ASSERTION !! (use sparingly!)
    // ═══════════════════════════════════════════════════════════════

    // !! converts nullable to non-null, throws NPE if null
    // val length3: Int = nullableName!!.length  // ❌ KotlinNullPointerException!

    val safeName: String = "Kotlin"
    val length3: Int = safeName!!.length  // ✅ OK — safeName is not null
    println("length3: $length3")

    // ═══════════════════════════════════════════════════════════════
    // 5. ELVIS OPERATOR ?: (default value)
    // ═══════════════════════════════════════════════════════════════

    // If left side is null, use right side
    val length4: Int = nullableName?.length ?: 0
    println("length4: $length4")  // 0

    val displayName: String = nullableName ?: "Unknown"
    println("displayName: $displayName")  // Unknown

    // Elvis with early return
    fun processName(name: String?): String {
        val nonNull: String = name ?: return "No name provided"
        return "Name: $nonNull"
    }
    println(processName(null))    // No name provided
    println(processName("Alice")) // Name: Alice

    // ═══════════════════════════════════════════════════════════════
    // 6. SAFE CASTS (as?)
    // ═══════════════════════════════════════════════════════════════

    val something: Any = "Hello"

    // Java: if (something instanceof String) { ... }
    // Kotlin: safe cast
    val safeString: String? = something as? String   // "Hello"
    val safeInt: Int? = something as? Int             // null (no exception!)

    println("safeString: $safeString")
    println("safeInt: $safeInt")

    // ═══════════════════════════════════════════════════════════════
    // 7. LET FUNCTION (scoped null check)
    // ═══════════════════════════════════════════════════════════════

    // let runs block only if not null
    nullableName?.let {
        println("Name is not null: $it")
    } ?: println("Name is null")

    // Java equivalent:
    // if (nullableName != null) {
    //     System.out.println("Name is not null: " + nullableName);
    // }

    // ═══════════════════════════════════════════════════════════════
    // 8. NULLABLE COLLECTIONS
    // ═══════════════════════════════════════════════════════════════

    // List<String?> — list of nullable strings
    val nullableList: List<String?> = listOf("A", null, "C")
    val nonNullList: List<String> = listOf("A", "B", "C")

    // List<String>? — nullable list itself
    val maybeNullList: List<String>? = null

    // ═══════════════════════════════════════════════════════════════
    // 9. FUNNEL OPERATOR ?.let + ?: combinations
    // ═══════════════════════════════════════════════════════════════

    fun greetUser(name: String?) {
        name?.let { nonNull ->
            println("Hello, $nonNull!")
        } ?: println("Hello, stranger!")
    }

    greetUser("Alice")  // Hello, Alice!
    greetUser(null)     // Hello, stranger!

    // ═══════════════════════════════════════════════════════════════
    // 10. PLATFORM TYPES (Java ↔ Kotlin interop)
    // ═══════════════════════════════════════════════════════════════

    // When Kotlin calls Java code, Java types become "platform types"
    // Platform types can be treated as nullable or non-nullable
    // This is where most Kotlin NPEs come from!

    // ═══════════════════════════════════════════════════════════════
    // 11. COMPARISON SUMMARY
    // ═══════════════════════════════════════════════════════════════
    //
    // JAVA (manual checks)          KOTLIN (type-safe)
    // ─────────────────────────────  ──────────────────────────────
    // if (s != null) s.length()      s?.length
    // s != null ? s.length : 0       s?.length ?: 0
    // if (s == null) throw ...       s!!  (rarely used)
    // if (s instanceof String) ...   s as? String
    // Optional.ofNullable(s)         s?.let { ... }
    // Objects.requireNonNull(s)      s ?: throw ...
    //
    // KEY RULES:
    // ─────────────────────────────────────────────────
    // 1. Default to non-null types (String, Int, List<T>)
    // 2. Use nullable types ONLY when null is a valid value
    // 3. Prefer ?. over !! (safe over unsafe)
    // 4. Use ?: for default values
    // 5. Use let for scoped operations on nullable
    // 6. Be careful with Java interop (platform types)
    //
    // ═══════════════════════════════════════════════════════════════
}
